File: 05-Algorithms/code/python/seb_alg/kruskal.py
"""Mirrors `typescript/src/kruskal.ts` exactly."""

from __future__ import annotations

from typing import Dict, List, Sequence, Tuple


class _DisjointSet:
    def __init__(self, n: int) -> None:
        self.parent = list(range(n))

    def find(self, x: int) -> int:
        while self.parent[x] != x:
            x = self.parent[x]
        return x

    def union(self, a: int, b: int) -> bool:
        ra, rb = self.find(a), self.find(b)
        if ra == rb:
            return False
        if ra < rb:
            self.parent[rb] = ra
        else:
            self.parent[ra] = rb
        return True


def kruskal(n: int, edges: Sequence[Sequence[int]]) -> Dict[str, object]:
    """Minimum spanning forest via Kruskal's algorithm: sort edges ascending
    by (weight, u, v), greedily add each edge connecting two different
    components. O(E log E).
    """
    sorted_edges: List[Tuple[int, int, int]] = sorted(
        ((u, v, w) for u, v, w in edges), key=lambda e: (e[2], e[0], e[1])
    )
    dsu = _DisjointSet(n)
    chosen: List[List[int]] = []
    weight = 0
    for u, v, w in sorted_edges:
        if dsu.union(u, v):
            chosen.append([u, v, w])
            weight += w
    return {"weight": weight, "edges": chosen}

File: 05-Algorithms/code/typescript/src/kruskal.ts
import { type WeightedEdge } from "./graphCommon.js";

export interface MstResult {
  weight: number;
  edges: WeightedEdge[];
}

class DisjointSet {
  private parent: number[];
  constructor(n: number) {
    this.parent = Array.from({ length: n }, (_, i) => i);
  }
  find(x: number): number {
    while (this.parent[x] !== x) x = this.parent[x];
    return x;
  }
  union(a: number, b: number): boolean {
    const ra = this.find(a);
    const rb = this.find(b);
    if (ra === rb) return false;
    // Union by lower-root-wins, not by rank: keeps the merge rule itself
    // deterministic and independent of insertion order, which matters here
    // only for reproducibility of internal state, not of the output (the
    // MST weight/edge set is unique up to weight ties handled below).
    if (ra < rb) this.parent[rb] = ra;
    else this.parent[ra] = rb;
    return true;
  }
}

/**
 * Minimum spanning forest via Kruskal's algorithm: sort all edges ascending
 * by (weight, u, v) -- the tie-break on endpoints makes the exact edge set
 * chosen deterministic even when multiple equal-weight MSTs exist -- then
 * greedily add each edge that connects two different components (checked
 * via union-find). Works on disconnected graphs (returns a minimum
 * spanning *forest*; `edges.length` will be less than n-1 in that case).
 * O(E log E).
 */
export function kruskal(n: number, edges: WeightedEdge[]): MstResult {
  const sorted = edges.slice().sort((a, b) => {
    if (a[2] !== b[2]) return a[2] - b[2];
    if (a[0] !== b[0]) return a[0] - b[0];
    return a[1] - b[1];
  });
  const dsu = new DisjointSet(n);
  const chosen: WeightedEdge[] = [];
  let weight = 0;
  for (const [u, v, w] of sorted) {
    if (dsu.union(u, v)) {
      chosen.push([u, v, w]);
      weight += w;
    }
  }
  return { weight, edges: chosen };
}
